src/ir.cpp

```diff
@@ -13125,6 +13125,10 @@ static IrInstruction *ir_analyze_array_cat(IrAnalyze *ira, IrInstructionBinOp *i
         ZigType *ptr_type = op1_type->data.structure.fields[slice_ptr_index].type_entry;
         child_type = ptr_type->data.pointer.child_type;
         ConstExprValue *ptr_val = &op1_val->data.x_struct.fields[slice_ptr_index];
+        if (ptr_val->data.x_ptr.special == ConstPtrSpecialInvalid) {
```

Comment:
The `ConstPtrSpecialInvalid` state isn't meant to be used in any meaningful way; it's more of a debug tool to detect when something is wrong. So this would actually have to be: `assert(ptr_val->data.x_ptr.special != ConstPtrSpecialInvalid)` and then the next step of debugging this issue would be to determine why the value is set to `ConstPtrSpecialInvalid` and not something valid. Note that when you use a debugger (gdb on linux, lldb on macos, lldb on windows) and it trips this assertion, the debugger will let you look at all the variables and look up the stack etc.

Looking around, I can see that this `assert` step has already been done; on the next line we see `assert(ptr_val->data.x_ptr.special == ConstPtrSpecialBaseArray);`. The fact that it is an `assert` means that the code relies on the assumption holding, and the fact that the assumption was violated, is the bug. So the way to solve this is to first understand why exactly this assumption has been invalidated. This will hopefully then lead to solving the root cause elsewhere.

One clue for working on ir.cpp is this gist: https://gist.github.com/andrewrk/5684434a2a8d4cbb08bb0d855c4f2ada
This makes the amount of code that gets analyzed only exactly what is in the example.zig file. So if you put your test case in it you can use those verbose command line options without it being overwhelming, and if you put a breakpoint somewhere in ir.cpp, it's likely to get hit (or not) based only on the contents of example.zig.

Hope that helps!

Reply:
Thanks @andrewrk. That helps a lot!